csharp/connection/TypeDBDatabase.cs

```diff
@@ -0,0 +1,196 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Linq;
+
+using com.vaticle.typedb.driver;
+using com.vaticle.typedb.driver.Api.Database;
+using com.vaticle.typedb.driver.Common;
+using com.vaticle.typedb.driver.Common.Exception;
+using DriverError = com.vaticle.typedb.driver.Common.Exception.Error.Driver;
+
+namespace com.vaticle.typedb.driver.Connection
+{
+    public class TypeDBDatabase : NativeObjectWrapper<pinvoke.Database>, IDatabase
+    {
+        public TypeDBDatabase(pinvoke.Database database)
+            : base(database)
+        {}
+
+        public string Name()
+        {
+            if (NativeObject == null || !NativeObject.IsOwned())
+            {
+                throw new TypeDBDriverException(DriverError.s_DatabaseDeleted);
+            }
+
+            return pinvoke.typedb_driver.database_get_name(NativeObject);
+        }
+
+        public string Schema()
+        {
+            if (NativeObject == null || !NativeObject.IsOwned())
+            {
+                throw new TypeDBDriverException(DriverError.s_DatabaseDeleted);
+            }
+
+            try
+            {
+                return pinvoke.typedb_driver.database_schema(NativeObject);
+            }
+            catch (pinvoke.Error e)
+            {
+                throw new TypeDBDriverException(e);
+            }
+        }
+
+        public string TypeSchema()
+        {
+            if (NativeObject == null || !NativeObject.IsOwned())
+            {
+                throw new TypeDBDriverException(DriverError.s_DatabaseDeleted);
+            }
+
+            try
+            {
+                return pinvoke.typedb_driver.database_type_schema(NativeObject);
+            }
+            catch (pinvoke.Error e)
+            {
+                throw new TypeDBDriverException(e);
+            }
+        }
+
+        public string RuleSchema()
+        {
+            if (NativeObject == null || !NativeObject.IsOwned())
+            {
+                throw new TypeDBDriverException(DriverError.s_DatabaseDeleted);
+            }
+
+            try
+            {
+                return pinvoke.typedb_driver.database_rule_schema(NativeObject);
+            }
+            catch (pinvoke.Error e)
+            {
+                throw new TypeDBDriverException(e);
+            }
+        }
+
+        public void Delete()
+        {
+            if (NativeObject == null || !NativeObject.IsOwned())
```

Comment:
IMO would be good to extract this kind of a thing out somewhere common so you can do `if (isValid(NativeObject)) { ... }`

Or even `validate(NativeObject, () -> DriverError.s_DatabaseDeleted)` which throws the error given if the validation fails in one line

Reply:
Sure, I was thinking about it, but decided to see how many similar checks I'll have in other files. I haven't fully cleaned up the code as I thought it would be mostly a review of the architecture, but thanks for the comments!

Reply:
Will return to it later, left a TODO.